SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp

```diff
@@ -0,0 +1,289 @@
+
+	return TUniquePtr<FArchiveLogFileReader>(new FArchiveLogFileReader(Handle, InFilename, Handle->Size(), BufferSize));
+}
+
+BEGIN_SLATE_FUNCTION_BUILD_OPTIMIZATION
+void SSpatialOutputLog::Construct(const FArguments& InArgs)
+{
+	SOutputLog::Construct(SOutputLog::FArguments());
+
+	// Remove ourselves as the constructor of our parent (SOutputLog) added 'this' as a remote output device.
+	GLog->RemoveOutputDevice(this);
+
+	LogReader = nullptr;
+
+	StartUpLogDirectoryWatcher(LocalDeploymentLogsDir);
+
+	// Set the LogReader to the latest launch.log if we can.
+	ReadLatestLogFile();
+}
+END_SLATE_FUNCTION_BUILD_OPTIMIZATION
+
+void SSpatialOutputLog::ReadLatestLogFile()
+{
+	FString NewestLogDir;
+	FDateTime NewestLogDirTime;
+
+	// Go through all log directories in the spatial logs and find the most recently created (if one exists) and print the log file to the Spatial Output.
+	bool GetNewestLogDir = IFileManager::Get().IterateDirectoryStat(*LocalDeploymentLogsDir, [&NewestLogDir, &NewestLogDirTime](const TCHAR* FileName, const FFileStatData& FileStats) {
+
+		if (FileStats.bIsDirectory)
+		{
+			if (FileStats.CreationTime > NewestLogDirTime)
+			{
+				NewestLogDir = FString(FileName);
+				NewestLogDirTime = FileStats.CreationTime;
+			}
+		}
+
+		return true;
+	});
+
+	if (GetNewestLogDir)
+	{
+		StartPollingLogFile(FPaths::Combine(NewestLogDir, LaunchLogFilename));
+	}
+}
+
+SSpatialOutputLog::~SSpatialOutputLog()
+{
+	CloseLogReader();
+
+	ShutdownLogDirectoryWatcher(LocalDeploymentLogsDir);
+
+	FCoreDelegates::OnHandleSystemError.RemoveAll(this);
+}
+
+void SSpatialOutputLog::OnCrash()
+{
+	CloseLogReader();
+
+	ShutdownLogDirectoryWatcher(LocalDeploymentLogsDir);
+}
+
+void SSpatialOutputLog::StartUpLogDirectoryWatcher(FString LogDirectory)
+{
+	AsyncTask(ENamedThreads::GameThread, [this, LogDirectory]
+	{
+		FDirectoryWatcherModule& DirectoryWatcherModule = FModuleManager::LoadModuleChecked<FDirectoryWatcherModule>(TEXT("DirectoryWatcher"));
+		if (IDirectoryWatcher * DirectoryWatcher = DirectoryWatcherModule.Get())
+		{
+			// Watch the log directory for changes.
+			if (FPaths::DirectoryExists(LogDirectory))
+			{
+				LogDirectoryChangedDelegate = IDirectoryWatcher::FDirectoryChanged::CreateRaw(this, &SSpatialOutputLog::OnLogDirectoryChanged);
+				DirectoryWatcher->RegisterDirectoryChangedCallback_Handle(LogDirectory, LogDirectoryChangedDelegate, LogDirectoryChangedDelegateHandle, IDirectoryWatcher::WatchOptions::IncludeDirectoryChanges | IDirectoryWatcher::WatchOptions::IgnoreChangesInSubtree);
+			}
+			else
+			{
+				UE_LOG(LogSpatialOutputLog, Error, TEXT("Local deployment log directory does not exist!"));
+			}
+		}
+	});
+}
+
+void SSpatialOutputLog::OnLogDirectoryChanged(const TArray<FFileChangeData>& FileChanges)
+{
+	// If this is a new folder creation then switch to watching the log files in that new log folder.
+	for (FFileChangeData FileChange : FileChanges)
+	{
+		if (FileChange.Action == FFileChangeData::FCA_Added)
```

Comment:
If you keep adding files while the editor is open does this cause us to poll each one? If so are they on the same timer?

Reply:
Ah so we try to close the reader each time.